### PR TITLE
[PIR] Optimized the matmul code mapping process

### PR DESCRIPTION
**Pull request by DrRyanHuang**
### PR types

Others

### PR changes

Others

### Description

- https://github.com/PaddlePaddle/Paddle/issues/56521
- https://github.com/PaddlePaddle/Paddle/pull/56550

### CODE REVIEW

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 1]
```diff
@@ -1279,76 +1315,32 @@ struct MulOpTranscriber : public OpTranscriber {
...
+        PADDLE_GET_CONST(int, op_desc.GetAttr("x_num_col_dims"));
+    const int y_num_col_dims =
+        PADDLE_GET_CONST(int, op_desc.GetAttr("y_num_col_dims"));
+
+    ValueInfo x_info =
+        getMatmulInfo(op_desc, op_desc.Input("X", true), param_map, "X");
+
+    const std::vector<int64_t>& x_shape = std::get<0>(x_info);
+    const dialect::DenseTensorType& x_tensor_type = std::get<1>(x_info);
+    pir::OpResult& x_value = std::get<2>(x_info);
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]:
```suggestion
    cosnt auto& [x_shape, x_tensor_type, x_value] = x_info;
```
Try structured binging  to simplify this.

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 2]
```diff
@@ -1279,76 +1315,32 @@ struct MulOpTranscriber : public OpTranscriber {
...
-    dialect::DenseTensorType y_tensor_type =
-        y_type.dyn_cast<dialect::DenseTensorType>();
-    std::vector<int64_t> y_shape = phi::vectorize(y_tensor_type.dims());
+    ValueInfo y_info =
+        getMatmulInfo(op_desc, op_desc.Input("Y", true), param_map, "Y");
+
+    const std::vector<int64_t>& y_shape = std::get<0>(y_info);
+    const dialect::DenseTensorType& y_tensor_type = std::get<1>(y_info);
+    pir::OpResult& y_value = std::get<2>(y_info);
+
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]: Same as above

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 3]
```diff
@@ -1733,96 +1582,61 @@ struct MulGradOpTranscriber : public OpTranscriber {
...
-               y_grad_type);
-    dialect::DenseTensorType y_grad_tensor_type =
-        y_grad_type.dyn_cast<dialect::DenseTensorType>();
-
-    VLOG(10) << "[" << op_desc.Type() << "] y_grad_shape change from "
-             << y_grad_tensor_type.dims() << " to " << dim_y;
-
-    param_map->PushValue(y_grad_var_name,
-                         VariableDefiningInfo(reshape_op_y.out(), false, -1));
+    gradReshape("Y@GRAD");
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]:
```suggestion
   if (y_grad_output.size()) {
      gradReshape("Y@GRAD");
   }
```
Keeping consistent form may be better.

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 4]
```diff
@@ -1556,115 +1475,45 @@ struct MulGradOpTranscriber : public OpTranscriber {
...
+        PADDLE_GET_CONST(int, op_desc.GetAttr("x_num_col_dims"));
+    const int y_num_col_dims =
+        PADDLE_GET_CONST(int, op_desc.GetAttr("y_num_col_dims"));
+
+    ValueInfo x_info =
+        getMatmulInfo(op_desc, op_desc.Input("X", true), param_map, "X");
+
+    const std::vector<int64_t>& x_shape = std::get<0>(x_info);
+    const dialect::DenseTensorType& x_tensor_type = std::get<1>(x_info);
+    pir::OpResult& x_value = std::get<2>(x_info);
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]: Same as above

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 5]
```diff
@@ -1556,115 +1475,45 @@ struct MulGradOpTranscriber : public OpTranscriber {
...
-               y_type);
-    dialect::DenseTensorType y_tensor_type =
-        y_type.dyn_cast<dialect::DenseTensorType>();
-    std::vector<int64_t> y_shape = phi::vectorize(y_tensor_type.dims());
+    ValueInfo y_info =
+        getMatmulInfo(op_desc, op_desc.Input("Y", true), param_map, "Y");
+
+    const std::vector<int64_t>& y_shape = std::get<0>(y_info);
+    const dialect::DenseTensorType& y_tensor_type = std::get<1>(y_info);
+    pir::OpResult& y_value = std::get<2>(y_info);
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]: Same as above

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 6]
```diff
@@ -1246,10 +1246,46 @@ struct TrilAndTriuOpTranscriber : public OpTranscriber {
   }
 };
 
+using ValueInfo =
+    std::tuple<std::vector<int64_t>, dialect::DenseTensorType, pir::OpResult>;
+
+ValueInfo getMatmulInfo(const OpDesc& op_desc,
```
Comment (kangguangli) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc]:
```suggestion
ValueInfo GetTensorInfoByVarName(const OpDesc& op_desc,
```

**Review by kangguangli — APPROVED** [review]

File: paddle/fluid/ir_adaptor/translator/op_translator.cc [hunk 7]
```diff
@@ -1733,96 +1574,63 @@ struct MulGradOpTranscriber : public OpTranscriber {
...
 
-      param_map->PushValue(x_grad_var_name,
-                           VariableDefiningInfo(reshape_op_x.out(), false, -1));
+    if (x_grad_output.size()) {
+      gradReshape("X@GRAD");
     }
 
     if (y_grad_output.size() < 1) {
       return;
     }
```
Comment (DrRyanHuang) [inline paddle/fluid/ir_adaptor/translator/op_translator.cc:1629]: 下一个pr删掉

**Pull request merged**